torch/csrc/jit/codegen/cuda/parser.cpp

```diff
@@ -49,7 +50,199 @@ const auto& boolAttr = Symbol::attr("profiled_bool");
+  return a <= b ? a : b;
+};
+
+class ValueHolder {
+ public:
+  // checks if given Val in target format exists.
+  bool hasValue(MemoryFormat format) const {
+    return vals_.count(format) != 0;
+  }
+
+  // returns Val in target format.
+  CgValue value(MemoryFormat format) const {
+    TORCH_INTERNAL_ASSERT(
+        hasValue(format), "accessing non existing c_last_value()");
+    return vals_.at(format);
+  }
+
+  // returns Val in target format if it exists, otherwise, transpose an existing
+  // copy and add that to bookkeeping.
+  CgValue maybeConvertValue(MemoryFormat format) {
+    if (hasValue(format)) {
+      return value(format);
+    }
+    // patching scalar value, because memory format doesn't carry real meaning.
+    if (!is_tensor_view_) {
+      return std::get<1>(getEntry());
+    }
+    auto [format_s, value_s] = getEntry();
+    auto val = convertValue(format, format_s, value_s);
+    vals_[format] = val;
+    return val;
+  }
+
+  // TODO: delete this and update accessor for value_map(_)
+  ValueHolder() {
+    TORCH_INTERNAL_ASSERT(false, "can't default constructor ValueHolder");
+  }
+
+  ValueHolder(CgValue val, MemoryFormat format = MemoryFormat::Contiguous) {
+    vals_[format] = val;
+    if (val->isA<TensorView>()) {
+      is_tensor_view_ = true;
+    }
+  }
+
+  // returns the MemoryFormat and codegen Val with the highest precedence among
+  // existing copies.
+  std::tuple<MemoryFormat, CgValue> getEntry() const {
+    static auto formats = {
+        MemoryFormat::ChannelsLast, MemoryFormat::Contiguous};
+    for (const auto& format : formats) {
+      if (hasValue(format)) {
+        return {format, value(format)};
+      }
+    }
+    TORCH_CHECK(false, "accessing empty ValueHolder");
+  }
+
+  // TODO: code cleaning in parser so we don't need these.
+  // returns Val*, keeping them here just so we have less code change.
+  CgValue operator*() const {
+    return std::get<1>(getEntry());
+  }
+  CgValue operator->() const {
+    return std::get<1>(getEntry());
+  }
+  operator CgValue() const {
+    return std::get<1>(getEntry());
+  }
+
+ private:
+  // helper function to convert value_s @ format_s to format_d
+  CgValue convertValue(
+      MemoryFormat format_d,
+      MemoryFormat format_s,
+      CgValue value_s) {
+    TORCH_INTERNAL_ASSERT(
+        value_s->isA<TensorView>(), "cannot convert non-TensorView");
+    auto tv = value_s->as<TensorView>();
+    CgValue value_d;
+    auto n_dim = tv->nDims();
+    switch (switch_pair(format_d, format_s)) {
+      case switch_pair(MemoryFormat::ChannelsLast, MemoryFormat::Contiguous): {
+        std::unordered_map<int, int> permutation_axes;
+        for (const auto i : c10::irange(n_dim - 2)) {
+          permutation_axes[n_dim - 1 - i] = n_dim - 2 - i;
+        }
+        permutation_axes[1] =
+            n_dim - 1; // {{n-1, n-2}, {n-2, n-3}, ... {1, n-1}}
+        value_d = transpose(tv, permutation_axes);
```

Comment:
Please let me know if segmenter has any problem with these IO transpose ops. I don't think it will have correctness issue as is but could need updating.

Reply:
That was a thing I struggled earlier. But per Christian's suggestion, we now move the permutation logic onto the vanilla `Fusion`, which is not segmented. So we should be good there.